engine/schema/src/com/cloud/vm/dao/UserVmDaoImpl.java

```diff
@@ -299,21 +304,32 @@ public void updateVM(long id, String displayName, boolean enable, Long osTypeId,
         return listBy(sc);
     }
 
-    @Override
-    public List<UserVmVO> listByNetworkIdAndStates(long networkId, State... states) {
-        if (UserVmSearch == null) {
+    /**
+     * Recreates UserVmSearch depending on network type, as nics on L2 networks have no ip addresses
+     * @param network network
+     */
+    private void recreateUserVmSeach(NetworkVO network) {
+        if (network != null) {
             SearchBuilder<NicVO> nicSearch = _nicDao.createSearchBuilder();
             nicSearch.and("networkId", nicSearch.entity().getNetworkId(), SearchCriteria.Op.EQ);
             nicSearch.and("removed", nicSearch.entity().getRemoved(), SearchCriteria.Op.NULL);
-            nicSearch.and().op("ip4Address", nicSearch.entity().getIPv4Address(), SearchCriteria.Op.NNULL);
-            nicSearch.or("ip6Address", nicSearch.entity().getIPv6Address(), SearchCriteria.Op.NNULL);
-            nicSearch.cp();
+            if (!network.getGuestType().equals(Network.GuestType.L2)) {
```

Comment:
Can the method `getGuestType` return `Null`?
If not, the code is ok. Otherwise, it might be more interesting to do `Network.GuestType.L2.equals(network.getGuestType())`

Or even using `==`, if it is an enum type, it is ok to use `==`